### PR TITLE
 Integrate nebula into the project

### PR DESCRIPTION
**Pull request by Warefoxe**
Integrate nebula into the project using cargo git reference. 
Create nebula plugin with javascript API

### CODE REVIEW

File: src-tauri/tauri-plugin-nebula/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    contexts: HashMap<String, Arc<Mutex<nebula::Context>>>,
+}
+
+#[derive(Default)]
+struct NebulaState {
+    models: Arc<Mutex<HashMap<String, NebulaModelState>>>,
+}
+
+#[tauri::command]
+async fn init_model<R: Runtime>(
```
Comment (ospfranco) [inline src-tauri/tauri-plugin-nebula/src/lib.rs:53]: Maybe some documentation on this functions would be useful?

File: src-tauri/tauri-plugin-nebula/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    async_runtime::Mutex,
+    plugin::{Builder, TauriPlugin},
+    AppHandle, Manager, Runtime, State,
+};
+
+use serde::{Serialize, Serializer};
+use base64::prelude::*;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
```
Comment (ospfranco) [inline src-tauri/tauri-plugin-nebula/src/lib.rs:14]: is it worth it to put on it's own file? will it be re-used as the plugin grows?

File: src-tauri/tauri-plugin-nebula/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+#[tauri::command]
+async fn model_context_eval_string<R: Runtime>(
+    model: String,
+    context: String,
+    data: String,
+    use_bos: bool,
+    _app: AppHandle<R>,
+    state: State<'_, NebulaState>,
+) -> Result<()> {
+    if let Some(mm) = state.models.lock().await.get_mut(&model) {
```
Comment (ospfranco) [inline src-tauri/tauri-plugin-nebula/src/lib.rs:151]: noob question, where do the models stream responses back to JS? don't quite understand what each function does but I'm not seeing any event emit.

File: src/widgets/chat/api/chatApi.ts [hunk 4]
```diff
@@ -1,7 +1,10 @@
+import { invoke } from '@tauri-apps/api';
 import OpenAI from 'openai';
 
 import { ChatMsg } from '@/db/chat';
 
+invoke('plugin:nebula|init');
```
Comment (ospfranco) [inline src/widgets/chat/api/chatApi.ts:6]: This can throw and is also async right?

File: src/widgets/chat/model/chat.ts [hunk 5]
```diff
@@ -1,240 +1,241 @@
...
-
-export const $streamedText = combine($streamedMsgId, $messages.data, (msgId, data) => {
-  const msg = data[msgId!];
-  return msg?.text || '';
-});
-
-export const { askQuestion, startNew: startNewChat, switch: switchChat } = chatEvt;
-
-createUserMsg.fail.watch((a) => console.log('askQuestion', a));
+import { combine, createEffect, createEvent, createStore, sample } from 'effector';
```
Comment (ospfranco) [inline src/widgets/chat/model/chat.ts:1]: @demetrxx can you review changes on the frontend?

File: src-tauri/tauri-plugin-nebula/src/lib.rs (hunk 2 above)
Comment (Warefoxe) [reply src-tauri/tauri-plugin-nebula/src/lib.rs:14]: Yes, I will made it

File: src-tauri/tauri-plugin-nebula/src/lib.rs (hunk 3 above)
Comment (Warefoxe) [reply src-tauri/tauri-plugin-nebula/src/lib.rs:151]: I suppose not use events, just call predict function and get predicted text, but if needed stream I will try to implement it

File: src/widgets/chat/api/chatApi.ts (hunk 4 above)
Comment (Warefoxe) [reply src/widgets/chat/api/chatApi.ts:6]: yes, will update this

**Comment by ospfranco** [comment]
@Warefoxe can you merge the tip of `develop` into this branch? It will run CI to make sure everything builds on all platforms.

**Comment by Warefoxe** [comment]
@ospfranco , I am change integration code, now it try to simulate stream with events, please check updated code.

**Comment by ospfranco** [comment]
Tests would also be good. Even if there is no model to use, you can just mock the responses, but make sure that the correct events are emitted and so.

File: src-tauri/tauri-plugin-nebula/src/lib.rs (hunk 1 above)
Comment (ospfranco) [reply src-tauri/tauri-plugin-nebula/src/lib.rs:37]: Docs are still missing

File: src/entities/model/index.ts [hunk 6]
```diff
@@ -1,5 +1,2 @@
 export { downloadModel } from './api/downloadModel.ts';
```
Comment (ospfranco) [inline src/entities/model/index.ts:1]: Barrel files are not good, they will degrade performance once the app is large enough. I would strongly discourage their use.

https://marvinh.dev/blog/speeding-up-javascript-ecosystem-part-7/

File: src/entities/model/lib/modelExists.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { BaseDirectory, readDir } from '@tauri-apps/api/fs';
+
+import { AIModelName, models, MODELS_DIR_NAME } from '@/constants';
+
+export async function modelExists(model?: AIModelName) {
+  const entries = await readDir(MODELS_DIR_NAME, { dir: BaseDirectory.AppLocalData });
```
Comment (demetrxx) [inline src/entities/model/lib/modelExists.ts:6]: According to Nick's comment, we need to change .AppLocalData to .appData

File: src/widgets/chat/api/chatApi.ts [hunk 8]
```diff
@@ -1,53 +1,60 @@
...
-export interface AIChatMessage {
-  role: 'user' | 'assistant';
-  content: string;
-}
-
-function getAIChatMessages(messages: ChatMsg[]): AIChatMessage[] {
-  return messages
-    .slice(0, -1)
-    .map((msg) => ({ role: msg.isUser ? 'user' : 'assistant', content: msg.text }));
-}
```
Comment (demetrxx) [inline src/widgets/chat/api/chatApi.ts:20]: 1. How do you distinguish between user and assistant messages then?
2. Why don't you cut last message, it's an empty assistant message for UI purposes only

File: src/entities/model/api/downloadModel.ts [hunk 9]
```diff
@@ -2,14 +2,14 @@ import { createDir } from '@tauri-apps/api/fs';
...
+import { AIModelName, models, MODELS_DIR_NAME } from '@/constants.ts';
+
 import { getModelPath } from '../lib/getModelPath.ts';
 
-export async function downloadModel(model: AIModel, handler: Parameters<typeof download>[2]) {
+export async function downloadModel(model: AIModelName, handler: Parameters<typeof download>[2]) {
   const { url } = models[model];
 
-  await createDir(MODELS_DIR, { dir: BaseDirectory.AppLocalData, recursive: true });
+  await createDir(MODELS_DIR_NAME, { dir: BaseDirectory.AppLocalData, recursive: true });
```
Comment (demetrxx) [inline src/entities/model/api/downloadModel.ts:12]: change to .appData

File: src/entities/model/lib/getModelPath.ts [hunk 10]
```diff
@@ -1,10 +1,16 @@
...
 import { appLocalDataDir, resolve } from '@tauri-apps/api/path';
 
-import { AIModel, models } from '../consts/models.ts';
-import { MODELS_DIR } from '../consts/modelsPath.ts';
+import { AIModelName, models, MODELS_DIR_NAME } from '@/constants';
 
-export async function getModelPath(model: AIModel) {
+export async function getModelPath(model: AIModelName) {
   const { localName } = models[model];
   const localDir = await appLocalDataDir();
```
Comment (demetrxx) [inline src/entities/model/lib/getModelPath.ts:8]: change to appDataDir()

**Review by demetrxx — CHANGES_REQUESTED** [review]

File: src/widgets/chat/api/chatApi.ts (hunk 8 above)
Comment (ospfranco) [reply src/widgets/chat/api/chatApi.ts:20]: 1. I don't know how this is supposed to work. I assume one should fetch the entire conversation into the model, it doesn't really care? @Warefoxe 
2. I will take a look into this

Comment (Warefoxe) [reply src/widgets/chat/api/chatApi.ts:20]: I have add parameter to context initialisation with set of messages

**Comment by ospfranco** [comment]
Tests are still missing but if @demetrxx says everything is ok, I will take care of them on a separate PR please.

**Review by demetrxx — APPROVED** [review]
Looks good to me!

**Pull request merged**